File: Leetcode/563.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    unordered_map<TreeNode*, int> node2sum;
    int tilt = 0;
    int dfs1(TreeNode* root) {
        if (root->left && root->right) {
            node2sum[root] = root->val + dfs1(root->left) + dfs1(root->right);
        }
        else if (!root->left && root->right) {
            node2sum[root] = root->val + dfs1(root->right);
        }
        else if (root->left && !root->right) {
            node2sum[root] = root->val + dfs1(root->left);
        }
        else {
            node2sum[root] = root->val;
        }
        return node2sum[root];
    }
    void dfs2(TreeNode* root) {
        if (root->left && root->right) {
            tilt += abs(node2sum[root->left] - node2sum[root->right]);
            dfs2(root->left);
            dfs2(root->right);
        }
        else if (!root->left && root->right) {
            tilt += abs(node2sum[root->right]);
            dfs2(root->right);
        }
        else if (root->left && !root->right) {
            tilt += abs(node2sum[root->left]);
            dfs2(root->left);
        }
    }
    int findTilt(TreeNode* root) {
        if (!root) return 0;
        int rootSum = dfs1(root);
        dfs2(root);
        return tilt;
    }
};

File: Leetcode/563/563.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
var dp map[*TreeNode]int

func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

func solve(node *TreeNode) int {
	_, ok := dp[node]
	if ok {
		return dp[node]
	}
	dp[node] = node.Val
	if node.Left != nil {
		dp[node] += solve(node.Left)
	}
	if node.Right != nil {
		dp[node] += solve(node.Right)
	}
	return dp[node]
}

func dfs(node *TreeNode, sum *int) {
	if node.Left != nil && node.Right != nil {
		*sum += abs(solve(node.Left) - solve(node.Right))
		dfs(node.Left, sum)
		dfs(node.Right, sum)
	} else if node.Left != nil {
		*sum += abs(solve(node.Left))
		dfs(node.Left, sum)
	} else if node.Right != nil {
		*sum += abs(solve(node.Right))
		dfs(node.Right, sum)
	}
}

func findTilt(root *TreeNode) int {
	if root == nil {
		return 0
	}
	dp = make(map[*TreeNode]int)
	sum := 0
	dfs(root, &sum)
	return sum
}